### PR TITLE
website/docs: add content about bindings

### PR DESCRIPTION
**Pull request by tanberry**
This PR:
1. adds a new topic about Bindings, with an overview page and a page that will list all of the How Tos (bind a binding to stage, to a policy, to a group, etc)
2. adds new content to the procedural for how to create an app in the Wizard, with info about the new Bindings option
3. add new content about App Entitlements
4. adds a lot of links between all the things, because Bindings are everywhere

TODO:
- [ ] in a future PR, add back in the link to [`entitlements` scope](../providers/oauth2/index.md#default--special-scopes) on line 51. (That link is broken for now until Jen's PR for app entitlements is merged.)

Related to https://github.com/goauthentik/authentik/pull/11587

-   [x] The documentation has been updated
-   [x] The documentation has been formatted (`make website`)

### CODE REVIEW

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 1]
```diff
@@ -26,7 +26,11 @@ To add an application to authentik and have it display on users' **My applicatio
 
 ## Authorization
 
-Application access can be configured using (Policy) bindings. Click on an application in the applications list, and select the _Policy / Group / User Bindings_ tab. There you can bind users/groups/policies to grant them access. When nothing is bound, everyone has access. You can use this to grant access to one or multiple users/groups, or dynamically give access using policies.
+Application access can be configured using either (Policy) bindings or Application Entitlements.
+
+### Policy-driven authorization
+
+To use a policy to control which users or groups can access an application, click on an application in the applications list, and select the **Policy/Group/User Bindings** tab. There you can bind users/groups/policies to grant them access. When nothing is bound, everyone has access. You can use this to grant access to one or multiple users/groups, or dynamically give access using policies.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:33]: Question: "When nothing is bound, everyone has access. You can use this to grant access..."

That confuses me. Aside from the antecedent issue (what does "this" refer to in that sentence?), if everyone has access, why would I want to grant _more_ access?  

> When nothing is bound, everyone has access.  Binding a policy restricts access to specifiec Users or Groups, or by other custom policies such as restriction to a set time-of-day or a geographic region.

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 2]
```diff
@@ -35,6 +39,18 @@ When multiple policies/groups/users are attached, you can configure the _Policy
 - Require users to pass all bindings/be member of all groups (ALL), or
 - Require users to pass either binding/be member of either group (ANY)
 
+### Application Entitlements
+
+Another method to control which users or groups can access an application is to create an Application Entitlement (which defines the specific application(s)), and then bind that to specific groups or users.
+
+1. To create an Application Entitlement open the Admin interface and navigate to **Applications -> Applications**.
+2. Click the **Application entitlements** tab, and then click **Create entitlement**. Provide a name for the entitlement, enter any optional **Attributes** , and then click **Create**.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:47]: I don't see this feature on my latest build of `main`.  Does it need to be enabled somewhere?

Do you navigate to a new tab on the ApplicationsList page, or do you navigate to a specific Application and bind it?  If the latter, there's a step missing between 1 and 2: "Choose an application."

**Review by kensternberg-authentik — COMMENTED** [review]
Had a couple of concerns.

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 1 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:33]: whoa, very cool examples of ways ot use policies to restrict access, thank you!

**Review by kensternberg-authentik — APPROVED** [review]
Much more clear, thanks!

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 3]
```diff
@@ -20,13 +20,25 @@ To add an application to authentik and have it display on users' **My applicatio
 
 2. Click **Create with Wizard**. (Alternatively, use our legacy process and click **Create**. The legacy process requires that the application and its authentication provider be configured separately.)
 
-3. In the **New application** wizard, define the application details, the provider type and configuration, and then click **Submit**.
+3. In the **New application** wizard, define the application details, the provider type, bindings for the application.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:23]: Suggestion: "... and optional bindings for the application."  This mirrors the "... and optional UI settings" used down on line 25.

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 4]
```diff
@@ -20,13 +20,25 @@ To add an application to authentik and have it display on users' **My applicatio
...
+- **Application**: provide a name, an optional group for the type of application, the policy engine mode, and optional UI settings.
+
+- **Choose a Provider**: Select the provider types for this application.
+
+- **Configure a Provider**: Provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and any additional required configurations.
+
+- **Configure Bindings**: To manage the display of the new application on the **My applications** page, you can optionally define [bindings](../flows-stages/bindings/index.md) for a specific policy, group, or user. To do so in the Wizard, click **Bind existing policy/group/user** to add a binding. You can select an existing policy binding, or create a new binding specifically for a group or user. For example, if you select **User** and then choose an existing user from the drop-down menu, you create a new binding between the user and this specific application. Note that if you do not define any bindings, then all users have access to the application. For more information, refer to [authorization](#authorization).
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:31]: I worry that this confuses the issue for experienced users.  I don't know that there's a need to explain the difference between Users, Groups, and other Policies here.  That's the problem with having "Policy" Policy bindings.  Maybe we should call them something else, I dunno, "Custom" bindings.

File: website/docs/add-secure-apps/flows-stages/bindings/index.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+---
+title: Bindings
+---
+
+A binding is, simply put, a connection between two components (a flow, stage, policy, user, or group) _that adds additional content_ to one those existing components.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:5]: "Content" is such a weird word.  "Functionality?" "Intelligence?"

File: website/docs/add-secure-apps/flows-stages/stages/index.md [hunk 6]
```diff
@@ -54,3 +54,24 @@ To bind a stage to a flow, follow these steps:
 3. In the list of flows, click the name of the flow to which you want to bind one or more stages.
 4. On the Flow page, click the **Stage Bindings** tab at the top.
 5. Here, you can decide if you want to create a new stage and bind it to the flow (**Create and bind Stage**), or if you want to select an existing stage and bind it to the flow (**Bind existing stage**).
+
+## Bind users and groups to a flow's stage binding
+
+You can use bindings to determine whehther or not a stage is presented to a single user or any users within a group. You do this by binding the user or group to a stage binding within a specific flow. For example, if you have a flow that contains a stage that prompts the user for multi-factor authentication, but you only want certain users to see this stage (and fulfill the MFA prompt), then you would bind the appropriate group (or single user) to the stage binding for that flow.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/stages/index.md:60]: So... what do the users who _don't_ see this stage experience?

**Review by kensternberg-authentik — APPROVED** [review]
I'm approving because I suspect I may misunderstand some of the instructions. But if I misunderstand, others may as well, so I left comments.

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 4 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:31]: It is a bit confusing, precisely because there is not an exact parallel between the three tab options of Policy/Group/User. They behave differently. With Policy, you have to select an existing policy bindings. With Group or User, you are choosing to create a new binding.

We have to document how the product works, but tell me where the wording could change to make it more clear @kensternberg-authentik ?

File: website/docs/add-secure-apps/flows-stages/bindings/index.md (hunk 5 above)
Comment (tanberry) [reply website/docs/add-secure-apps/flows-stages/bindings/index.md:5]: I agree, content is not at all the right word here. Let me try to reword...

File: website/docs/add-secure-apps/flows-stages/stages/index.md (hunk 6 above)
Comment (tanberry) [reply website/docs/add-secure-apps/flows-stages/stages/index.md:60]: They just see the next stage (that they are allowed to see) in the flow. @BeryJu please confirm.

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 4 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:31]: I am interested in the idea of calling them something else... maybe Application bindings. And then once you say you want an application binding, you choose via radio buttons if you want to bind a Group, a User, or an existing Policy binding. So it ends up with same result, but the beginning part, with the three tabs, might not be so confusing.

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 7]
```diff
@@ -20,13 +20,21 @@ To add an application to authentik and have it display on users' **My applicatio
 
 2. Click **Create with Wizard**. (Alternatively, use our legacy process and click **Create**. The legacy process requires that the application and its authentication provider be configured separately.)
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:21]: Not part of this PR, but would it providing a link to the "legacy process" in its entirety? We have that as a User Story, right?  [Worried Princess Amadala Meme.]. Right?

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 8]
```diff
@@ -20,13 +20,21 @@ To add an application to authentik and have it display on users' **My applicatio
...
 
-## Authorization
+    - **Choose a Provider**: Select the provider types for this application.
 
-Application access can be configured using (Policy) bindings. Click on an application in the applications list, and select the _Policy / Group / User Bindings_ tab. There you can bind users/groups/policies to grant them access. When nothing is bound, everyone has access. You can use this to grant access to one or multiple users/groups, or dynamically give access using policies.
+    - **Configure a Provider**: Provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and any additional required configurations.
+
+    - **Configure Bindings**: To manage the display of the new application on the **My applications** page, you can optionally create a [binding](../flows-stages/bindings/index.md) between the application and a specific policy, group, or user. Note that if you do not define any bindings, then all users have access to the application. For more information about user access, refer to our documentation about [authorization](#policy-driven-authorization) and [hiding an application](#hide-applications).
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:31]: Suggestion "... to manage the listing and access to applications on a user's **My Applications** page..."

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 9]
```diff
@@ -20,13 +20,21 @@ To add an application to authentik and have it display on users' **My applicatio
...
+
+4. On the **Review and Submit Application** panel, review the configuration for the new application and its provider, and then click **Submit**.
+
+## Policy-driven authorization
+
+To use a [policy](../../customize/policies/index.md) to control which users or groups can access an application, click on an application in the applications list, and select the **Policy/Group/User Bindings** tab. There you can bind users/groups/policies to grant them access. When nothing is bound, everyone has access. Binding a policy restricts access to specific Users or Groups, or by other custom policies such as restriction to a set time-of-day or a geographic region.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:37]: Nit: No comma needed in "... click on an application in the applications list and select..."  It's a two-step, the Oxford comma is dead-horse levels of overkill here.

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 10]
```diff
@@ -35,6 +43,49 @@ When multiple policies/groups/users are attached, you can configure the _Policy
 - Require users to pass all bindings/be member of all groups (ALL), or
 - Require users to pass either binding/be member of either group (ANY)
 
+## Application Entitlements
+
+<span class="badge badge--preview">Preview</span>
+<span class="badge badge--version">authentik 2024.12+</span>
+
+Application entitlements can be used through authentik to manage authorization within an application (what areas of the app can users or groups can access). Entitlements are scoped to a single application and can be bound to multiple users/groups (binding policies is not currently supported), giving them access to the entitlement. An application can either check for the name of the entitlement (via the `entitlements` scope), or via attributes stored in entitlements.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:51]: Typo: "... app users can..." instead of "... app can users..."

*Dyslexics of the world, untie!*

Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:51]: Suggestion: "... bound to multiple users and/or groups."  The current wording makes it sound like "users/groups" is some kind of object, concept, or entity in our system.

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 11]
```diff
@@ -35,6 +43,49 @@ When multiple policies/groups/users are attached, you can configure the _Policy
...
+
+An authentik admin can create an entitlement [in the Admin interface](#create-an-application-entitlement) or using the [authentik API](../../developer-docs/api/api.md).
+
+Because entitlements exist within an application, names of entitlements must be unique within an application. This also means that entitlements are deleted when an application is deleted.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:55]: Suggestion: Wordy.  "Entitlements associated with an application will be deleted when the application is deleted."
Alternatively, "Entitlements bound to an application..."

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 12]
```diff
@@ -35,6 +43,49 @@ When multiple policies/groups/users are attached, you can configure the _Policy
...
+
+### Using entitlements
+
+Entitlements a user has access to can be retrieved using the `user.app_entitlements()` function in property mappings/policies. This function needs to be passed the specific application for which to get the entitlements. For example:
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:59]: Nit: propositional ending.  "Entitlements to which a user has access can be retrieved" or "Entitlements granted to a user can be retrieved..."

Shouldn't that be "property mappings → policies"?

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 13]
```diff
@@ -35,6 +43,49 @@ When multiple policies/groups/users are attached, you can configure the _Policy
...
+```python
+attrs = request.user.app_entitlements(provider.application)
+return {
+    "my_attr": attrs.get("my_attr")
+}
+```
+
+### Create an application entitlement
+
+1. To create an application entitlement open the Admin interface and navigate to **Applications -> Applications**.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:81]: Suggestion: "To create an application entitlement:" might work better as a phrase appearing _above_ the list, since it's the topic of this list of steps, not the topic of Item 1 or whatever.

```
To create an application entitlement:

1. Open the Admin Interface and navigate to **Applications → Applications**...
```

File: website/docs/add-secure-apps/applications/manage_apps.md [hunk 14]
```diff
@@ -35,6 +43,49 @@ When multiple policies/groups/users are attached, you can configure the _Policy
...
+2. Click the name of the application for which you want to create an entitlement.
+3. Click the **Application entitlements** tab at the top of the page, and then click **Create entitlement**. Provide a name for the entitlement, enter any optional **Attributes**, and then click **Create**.
+4. Locate the entitlement to which you want to bind a user or group, and then **click the caret (>) to expand the entitlement details.**
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/applications/manage_apps.md:84]: Where would I go to "locate the entitlement?"

File: website/docs/add-secure-apps/flows-stages/bindings/index.md [hunk 15]
```diff
@@ -0,0 +1,33 @@
+---
+title: Bindings
+---
+
+A binding is, simply put, a connection between two components (a flow, stage, policy, user, or group). The use of the binding adds additional functionality to one those existing components; for example, a policy binding can cause a new stage to be presented within a flow, for a specific user.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:5]: Typo and suggestion: "The use of a binding can add additional functionality to an existing component..."

Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:5]: Typo and suggestion: "... presented within a flow to a specific user or group."

File: website/docs/add-secure-apps/flows-stages/bindings/index.md [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+- stage bindings
+- policy bindings
+- user and group bindings
+
+A _stage binding_ connects a stage to a flow. The "additional content" (i.e. the content in the stage) is now added to the flow.
+
+A _policy binding_ connects a specific policy to a flow or to a stage. With the binding, the flow (or stage) will now have additional content (i.e. the policy rules).
+
+You can also bind groups and users to another component (a policy, a stage, a flow, etc.). For example, you can create a binding for a specific group, and then [bind that to a stage binding](../stages/index.md#bind-users-and-groups-to-a-flows-stage-binding), with the result that everyone in that group now will see that stage (and any policies bound to that stage) as part of their flow. Or more specifically, and going one step deeper, you can also _bind a binding to a binding_.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:25]: As always, I wonder how to describe what happens for people *not* in the group.

File: website/docs/add-secure-apps/flows-stages/bindings/index.md [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+Bindings are also used for [Application Entitlements](../../applications/manage_apps.md#application-entitlements), where you can bind specific users or groups to an application as a way to manage who has access to the application.
+
+It's important to remember that bindings are instantiated objects themselves, and conceptually can be considered as the "connector" between two components. This is why you might read about "binding a binding", because technically, a binding is "spliced" into another binding, in order to intercept and enforce the criteria defined in the second binding.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:29]: Suggestion: "... as _a_ 'connector'..."  I suggest we avoid the idea that bindings are singular, that you can only have one enforcement or modifying object per object-to-object relationship.

File: website/docs/add-secure-apps/flows-stages/bindings/index.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+:::info
+Be aware that some stages and flows do not allow user or group bindings, because in certain scenarios (authentication or enrollment), the flow plan doesn't yet know who the user or group is.
```
Comment (kensternberg-authentik) [inline website/docs/add-secure-apps/flows-stages/bindings/index.md:32]: Suggestion: "In scenarios where the user or group cannot be known, such as authentication or enrollment, user and group bindings will not work. To prevent mistakes, stages and flows meant to be seen by unauthorized or pre-authorized users will not permit user and group bindings." (Is this true?)

File: website/docs/customize/brands.md [hunk 19]
```diff
@@ -9,7 +9,7 @@ The main settings that brands influence are flows and branding.
 
 ## Flows
 
-You can explicitly select, in your instance's Brand settings, the default flow to use for the following configurations:
+You can explicitly select, in your instance's Brand settings, the _default flows_ to use for the brand. To do so, log in as an administrator, open the Admin interface, and navigate to **System -> Brands**. There you can optionally configure these default flows:
```
Comment (kensternberg-authentik) [inline website/docs/customize/brands.md:12]: Suggestion: "... that brand..." or "... the selected brand..." or "... the current brand..."

**Review by kensternberg-authentik — COMMENTED** [review]
Man, I'm in a nitpicky mood.  Sorry about this.  I still learned a lot just from reading it.

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 7 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:21]: Sorry to the Princess, but I don't think we ever documenting the soon-to-be-legacy way. I was holding out for the Wizard. (But in principle I agree!)

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 10 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:51]: Yes, agree, we have quite a few places in Docs where the forward slash does a lot of work. Thanks for catching this one.

File: website/docs/add-secure-apps/flows-stages/bindings/index.md (hunk 16 above)
Comment (tanberry) [reply website/docs/add-secure-apps/flows-stages/bindings/index.md:25]: You're too caring. They get nothing. Kidding. Actually, serious.

File: website/docs/add-secure-apps/flows-stages/bindings/index.md (hunk 17 above)
Comment (tanberry) [reply website/docs/add-secure-apps/flows-stages/bindings/index.md:29]: nice catch, nice subtly

File: website/docs/add-secure-apps/flows-stages/bindings/index.md (hunk 18 above)
Comment (tanberry) [reply website/docs/add-secure-apps/flows-stages/bindings/index.md:32]: I'm not actually sure. The original sentence came from Jens telling me that before the system nows who a user is, they can't possibly create a bindings between the stage or flow and the [unknown] user. I think it is that simple, but will find out. So I don;t think it is as fancy as your second sentence... it's not that the system is explicitly preventing because the user is not supposed to see it (unauthorized to see it) but rather that the user is not yet authenticated, so we don't know what user object to bind too. @BeryJu?

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 9 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:37]: Ahah, you made me miss something that my BMC Editor would have had fits about.. for final step one has to say ", and then click..."

File: website/docs/add-secure-apps/applications/manage_apps.md (hunk 14 above)
Comment (tanberry) [reply website/docs/add-secure-apps/applications/manage_apps.md:84]: Ah... in the list, but i will add that.
